Makefile

```diff
@@ -0,0 +1,103 @@
+
+PYTHON := python
+
+SHELL_ENV := .shell-env
+
+SHELL_ENV_EXISTS = $(SHELL_ENV)/.exists
+$(SHELL_ENV_EXISTS):
+	@mkdir -p $(shell dirname $@)
+	@touch $@
+
+PYTHON_VIRTUALENV_SRC = $(SHELL_ENV)/$(shell basename $(VIRTUALENV_SRC))
+$(PYTHON_VIRTUALENV_SRC): $(SHELL_ENV_EXISTS)
+	@rm -f $@.tmp
+	@$(PYTHON) hack/download.py $(VIRTUALENV_SRC) $@.tmp
+	@echo $(VIRTUALENV_SRC_SHA256SUM)  $@.tmp | sha256sum -c > /dev/null
+	@mv $@.tmp $@
+
+PYTHON_VIRTUALENV = $(SHELL_ENV)/$(shell basename $(PYTHON_VIRTUALENV_SRC) .tar.gz)
+PYTHON_VIRTUALENV_EXISTS = $(PYTHON_VIRTUALENV)/.exists
+$(PYTHON_VIRTUALENV_EXISTS): $(SHELL_ENV_EXISTS) $(PYTHON_VIRTUALENV_SRC)
+	@rm -rf $(PYTHON_VIRTUALENV)
+	@tar -xf $(PYTHON_VIRTUALENV_SRC) -C $(shell dirname $(PYTHON_VIRTUALENV))
+	@touch $@
+
+VENV = $(SHELL_ENV)/metal-k8s
+VENV_BIN = $(VENV)/bin
+VENV_ACTIVATE = $(VENV_BIN)/activate
+
+VENV_EXISTS = $(VENV)/.exists
+$(VENV_EXISTS): $(PYTHON_VIRTUALENV_EXISTS)
+	@rm -rf $(VENV)
+	@echo "Creating virtualenv..."
+	@PYTHONPATH=$(PYTHON_VIRTUALENV) $(PYTHON) -m virtualenv $(VENV) > /dev/null
+	@touch $@
+
+REQUIREMENTS_INSTALLED = $(SHELL_ENV)/.requirements_installed
+$(REQUIREMENTS_INSTALLED): $(VENV_EXISTS) requirements.txt
+	@echo "Installing Python dependencies..."
+	@source $(VENV_ACTIVATE) && pip install -r requirements.txt > /dev/null
+	@touch $@
+
+KUBECTL_BIN = $(SHELL_ENV)/kubectl-$(KUBECTL_VERSION)
+$(KUBECTL_BIN): $(SHELL_ENV_EXISTS)
+	@rm -f $@.tmp
+	@echo "Downloading kubectl..."
+	@$(PYTHON) hack/download.py $(KUBECTL_SRC) $@.tmp
+	@echo $(KUBECTL_SRC_SHA256SUM)  $@.tmp | sha256sum -c > /dev/null
+	@chmod a+x $@.tmp
+	@mv $@.tmp $@
+
+KUBECTL = $(VENV_BIN)/kubectl
+$(KUBECTL): $(KUBECTL_BIN) $(VENV_EXISTS)
+	@ln -sf $(shell realpath $(KUBECTL_BIN)) $@
+
+HELM_SRC_TAR = $(SHELL_ENV)/$(shell basename $HELM_SRC)
+$(HELM_SRC_TAR): $(SHELL_ENV_EXISTS)
+	@rm -f $@.tmp
+	@echo "Downloading Helm..."
+	@$(PYTHON) hack/download.py $(HELM_SRC) $@.tmp
+	@echo $(HELM_SRC_SHA256SUM)  $@.tmp | sha256sum -c > /dev/null
+	@mv $@.tmp $@
+
+HELM_BIN = $(SHELL_ENV)/helm-$(HELM_VERSION)
+$(HELM_BIN): $(HELM_SRC_TAR) $(SHELL_ENV_EXISTS)
+	@rm -f $(SHELL_ENV)/$(UNAME_KERNEL)-$(UNAME_MACHINE)/helm
+	@tar -xf $(HELM_SRC_TAR) -C $(SHELL_ENV) $(UNAME_KERNEL)-$(UNAME_MACHINE)/helm
+	@echo $(HELM_BIN_SHA256SUM)  $(SHELL_ENV)/$(UNAME_KERNEL)-$(UNAME_MACHINE)/helm | sha256sum -c > /dev/null
+	@mv $(SHELL_ENV)/$(UNAME_KERNEL)-$(UNAME_MACHINE)/helm $@
+
+HELM = $(VENV_BIN)/helm
+$(HELM): $(HELM_BIN) $(VENV_EXISTS)
+	@ln -sf $(shell realpath $(HELM_BIN)) $@
+
+BASHRC = $(SHELL_ENV)/bashrc
+$(BASHRC): $(SHELL_ENV_EXISTS) hack/shell-bashrc
+	@rm -f $@
+	@sed s:@VENV_ACTIVATE@:$(VENV_ACTIVATE):g < hack/shell-bashrc > $@ || (rm -f $@; exit 1)
+
+shell: $(VENV_EXISTS) $(REQUIREMENTS_INSTALLED) $(KUBECTL) $(HELM) $(BASHRC)
+	@echo "Launching metal-k8s shell environment. Run 'exit' to quit."
+	@bash --rcfile $(BASHRC)
```

Comment:
Need to add a ||: so if the last command has an error exit code, it doesn't display 

```
(metal-k8s) [ballot@OnePower metal-k8s (tox-shell *=)]
$ exit
exit
Makefile:97: recipe for target 'shell' failed
make: *** [shell] Error 2
```

Reply:
I considered that before, but wasn't sure 'hiding' shell failures was a good approach. Fair enough, done.